### PR TITLE
Add support for bytes regular expressions

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR adds support for using `bytes` values as regular expressions to be used
with other `bytes` values.

Summary:
- Makes `regexp` generic on `type exprType` which can be `string` or `bytes`
   
  Currently, this PR doesn't assign a default to `exprType`. But we might want to
  make `string` the default type. See #14693 

- Removes `stringPart` from Regexp module. This was based on some comments in
  the code that says it is not tested, and in some places, support for it is not
  fully implemented.

  This is not absolutely necessary for this PR to move on and I can revert back.
  But it makes the interface and sometimes implementation much easier to remove
  that.

- Remove the deprecated `compile` with `out err`, deprecates the `compile`
  with `utf8`  argument and all lowercase arguments.

  The new `compile` uses `utf8` encoding with strings and `Latin1` with bytes
  internally.

- Adds `bytes` methods with regexp arguments. `bytes.this(reMatch)`,
  `bytes.search(regexp(bytes))` etc
  
- Adds support for doing formatted reads for `regexp(bytes)`.

  The internal class `channel_regexp_info` now stores only `bytes`. They are
  `decode`d into strings as necessary when they are extracted.

  Making this also generic cannot be easily done, I think. Because `readf`
  needs to create a `channel_regexp_info` instance before starting to read the
  arguments. And before starting to read the arguments we don't know whether
  they are `regexp(bytes)` or `regexp(string)`.

  Note that the type of the format string does not determine the type of the
  regular expression. Therefore, one can capture a bytes-based regular
  expression using `"%/*/"` and a string-based regular expression using
  `b"%/*/"`

- Modifies some tests to test bytes-based regular expressions in the same way
  string-based ones are tested.

- Adjusts the `RecordParser` module to use `regexp(string)`

- Adjusts mason to use new `compile`

- Adds tests for non-UTF8 regexp matching and casts.

- Minor adjustment in the docs to emphasize that we have raw string/bytes
  literals.

Test:
- [x] standard
- [x] gasnet

### CODE REVIEW

File: modules/standard/Regexp.chpl [hunk 1]
```diff
@@ -578,15 +572,19 @@ proc string.this(m:reMatch) {
   */
 pragma "ignore noinit"
 record regexp {
+
+  pragma "no doc"
+  type exprType;
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/standard/Regexp.chpl]: See https://github.com/chapel-lang/chapel/issues/14693

File: modules/standard/Regexp.chpl [hunk 2]
```diff
@@ -1046,20 +1033,24 @@ proc =(ref ret:regexp, x:regexp)
...
+      }
+    else
+      pattern = createBytesWithOwnedBuffer(cs);
   }
   return pattern;
 }
 // Cast string to regexp
 pragma "no doc"
-inline proc _cast(type t, x: string) throws where t == regexp {
+inline proc _cast(type t, x: ?valType) throws
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/standard/Regexp.chpl]: See https://github.com/chapel-lang/chapel/issues/14694

**Comment by e-kayrakli** [comment]
@mppf I think this is ready for another round of review. Maybe you can check the commits after your initial review one-by-one? Thanks!

File: modules/standard/Regexp.chpl [hunk 3]
```diff
@@ -854,9 +834,12 @@ record regexp {
 
       if pos < splitstart {
         // Yield splitted value
-        yield text[pos+1..splitstart];
+        if exprType == string then
+          yield text[pos+1..splitstart];
+        else 
+          yield text[(pos+1):int..splitstart:int];
```
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/Regexp.chpl]: Can you make an issue about this one too? Thanks.

File: modules/standard/Regexp.chpl [hunk 4]
```diff
@@ -408,37 +411,34 @@ class BadRegexpError : Error {
...
-   will halt with an error message.
+   Compile a regular expression. This routine will throw a
+   class:`BadRegexpError` if compilation failed.
 
-   :arg pattern: the string regular expression to compile.
-                 See :ref:`regular-expression-syntax` for details. Note that
-                 you may have to escape backslashes. For example, to
+   :arg pattern: the regular expression to compile. This argument can be string
+                 or bytes. See :ref:`regular-expression-syntax` for details.
+                 Note that you may have to escape backslashes. For example, to
```
Comment (mppf) [inline modules/standard/Regexp.chpl]: I know this is unrelated to your PR, but while updating this documentation could you please emphasize in it that raw string literals exist? The docs here predate the raw string literals and regexps are an improtant motivating example.

File: test/regexp/bytes/nonUTFIO.chpl [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  w.write("Pattern:Capture\n");
+  w.write(b"Patt\xffern:Capture\n");
+  w.write(b"Patt\xffern:Cap\xffture\n");
+  w.write(b"Pattern:Cap\xffture\n");
+  w.close();
+
+  // read from it -- following four reads are how these should be read
+  var r = f.reader();
+  var captureString: string;
+  var captureBytes: bytes;
```
Comment (mppf) [inline test/regexp/bytes/nonUTFIO.chpl]: Can you add some testing with passing a compiled bytes/string regexp to this test?

File: modules/standard/Regexp.chpl (hunk 3 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/standard/Regexp.chpl]: Here is a PR: https://github.com/chapel-lang/chapel/pull/14695

File: test/regexp/bytes/nonUTFIO.chpl (hunk 5 above)
Comment (e-kayrakli) [reply test/regexp/bytes/nonUTFIO.chpl]: Done. Current version of this test does that via the testRead helper.

**Pull request merged**